### ISSUE TITLE
CPP for EV3 Learning Doc

### ISSUE THREAD

**Issue by Tom-Shen** [moriarty/ros-ev3#6]
Sir：

Can you tell me where can I find Doc to learn CPP for EV3, I know how to use CPP, I just want to know how to operate device for EV3 by CPP?

Thanks!

**Comment by moriarty** [comment]
http://ev3dev-lang.readthedocs.org/en/latest/

**Comment by moriarty** [comment]
You can also use Python.

I usually test first with Python to avoid compiling on the ev3.

**Comment by moriarty** [comment]
The above link isn't the best for learning.

The ev3dev issues have a few examples where people have asked similar questions.

https://github.com/ev3dev/ev3dev/issues/584

I'm on my mobile so I can't easily search for any decent issues which are related but somewhere there is a long discussion with lots of helpful tips.

**Comment by Tom-Shen** [comment]
Thanks a lot，I will look at it, thank you for your very kind help!

**Comment by jabrena** [comment]
Hi @Tom-Shen,

If you like C++, take a look this API:
https://github.com/ddemidov/ev3dev-lang-cpp

Juan Antonio

**Comment by Tom-Shen** [comment]
Hello, @jabrena I have looked at https://github.com/ddemidov/ev3dev-lang-cpp, I am quite new to Linux CPP, I used to use MS VC and Mac Xcode IDE, it looks very complicated, I want to be simple at beginning, could you help me on this, I want to take the program "button-test.cpp" as a test, could you give me a simple Cmake file just include necessary stuff , so I can use it to build only one program at one time.
Very appreciate your help!
Thanks.

**Comment by Tom-Shen** [comment]
@moriarty If you can help? very appreciated!

**Comment by moriarty** [comment]
Hi,

I can try to take a look after work today. I'm on campus and my EV3 is at home.

But here is another link to take a look at: http://www.ev3dev.org/docs/tutorials/using-brickstrap-to-cross-compile/

It has a helloworld example in c. But you can use the same gcc command on a button-test.cpp or motor-test.cpp

**Comment by moriarty** [comment]
@Tom-Shen sorry, I must have forgotten to include installing the https://github.com/ddemidov/ev3dev-lang-cpp in my instructions. 

If you were able to follow my instructions to install the `ros_comm` bindings, then you have a working brickstrap environment, which is mentioned on the ev3dev-lang-cpp readme. 

I can try to add better instructions soon.

**Comment by moriarty** [comment]
I don't think it was that I forgot to include the instructions, but I think it was due to a change in how ev3dev-lang repository is structured, and I hadn't tested the c++ bindings since. 
In the past it contained all of the language bindings, I just tried on my ev3 which I hadn't updated in a long time and it was working. 

Quickest way to get up and running for now, note these instructions aren't including the ROS stuff, just how to compile c++ for the ev3, and copy the executable to the ev3. 
1. Start with step 3 from the ros_comm [install instructions](https://github.com/moriarty/ros-ev3/blob/master/brickstrap-build-status.md), Enter the brickstrap shell. This command should be run from where you created the brickstrap environment the first time. 
   
   ```
   user@host$ brickstrap -b ev3-ev3dev-jessie -d ev3dev-ros shell
   (brickstrap)root@host#
   ```
2. Install git, and clone ev3dev-lang-cpp:
   
   ```
   (brickstrap)root@host# apt-get install git
   (brickstrap)root@host# git clone https://github.com/ddemidov/ev3dev-lang-cpp.git
   (brickstrap)root@host# cd ev3dev-lang-cpp
   ```
3. From ev3dev-lang-cpp README.md, build project:
   
   ```
   (brickstrap)root@host# mkdir build && cd build
   (brickstrap)root@host# cmake .. -DEV3DEV_PLATFORM=EV3
   (brickstrap)root@host# make
   ```
   
   Change the `-DEV3DEV_PLATFORM=` if you use a brickpi.
   
   The commands above use the CMakeLists.txt from the ev3dev-lang-cpp project, which creates the static library `libev3dev.a` and defines the cmake function `add_ev3_executable`. Because the linking is done against a static library, the result could just be copied over to the ev3 using either a gui like filezilla or command line `scp`
4. Copy the executable, replacing the ip address with your ev3. 
   
   ```
   (brickstrap)root@host# scp demos/button-test robot@192.168.88.181:button-test
   ```
   
   The executables from the ev3dev-lang-cpp project will be in build, the `scp` command above copies the file directly to the robot user home directory.     

I didn't have a chance to get a quick test with ROS at the moment. But hopefully this will get you on the correct track.
